### PR TITLE
Module 4 async redux

### PR DESCRIPTION
**Pull request by jurica-zecevic**

### CODE REVIEW

File: src/components/CourseForm/CourseForm.jsx [hunk 1]
```diff
@@ -129,7 +157,7 @@ const CreateCourse = () => {
 							borderColor: !formValid.isDescriptionValid && 'var(--color-red)',
 						}}
 						placeholder='Enter description...'
-						value={formValues.description}
+						value={formValues.description || ''}
```
Comment (evasimonyi) [inline src/components/CourseForm/CourseForm.jsx:160]: why are the || '' changes necessary? formValues are by default empty strings, if action is update, then here you already set values:

if (courseToUpdate) {
				setFormValues({
					title: courseToUpdate.title,
					description: courseToUpdate.description,
					duration: courseToUpdate.duration,
				});

so not sure why is this necessary, please help me out:)

File: src/components/Courses/Courses.jsx [hunk 2]
```diff
@@ -62,7 +79,7 @@ const Courses = ({ coursesList, authorsList }) => {
 					</ul>
 				</section>
 			) : (
-				<EmptyCourseList userRole={userRole} />
+				<EmptyCourseList role={role} />
```
Comment (evasimonyi) [inline src/components/Courses/Courses.jsx:82]: in my opinion unnecessary change, userRole={role} would be ok, but matter of preference

File: src/components/CourseForm/CourseForm.jsx [hunk 3]
```diff
@@ -1,21 +1,22 @@
...
 import AuthorItem from './components/AuthorItem';
 import Button from '../../common/Button/Button';
 import Input from '../../common/Input/Input';
 import Textarea from '../../common/Textarea/Textarea';
 
-import styles from './CreateCourse.module.css';
+import styles from './CourseForm.module.css';
 
-const CreateCourse = () => {
+const CourseForm = ({ coursesList, authorsList }) => {
```
Comment (evasimonyi) [inline src/components/CourseForm/CourseForm.jsx:19]: why isn't this component access these from store itself? one of the advantages of redux is to eliminate passing props

File: src/components/CourseForm/CourseForm.jsx [hunk 4]
```diff
@@ -30,9 +31,32 @@ const CreateCourse = () => {
...
 
+	const { action, courseId } = useParams();
+	const isUpdate = action === 'update';
+
 	const dispatch = useDispatch();
 	const navigateCourses = useNavigate();
 
+	useEffect(() => {
+		if (isUpdate) {
+			const courseToUpdate = coursesList.find(
```
Comment (evasimonyi) [inline src/components/CourseForm/CourseForm.jsx:42]: this calculation doesn't have to run if `authorsList` is updated, but currently it does, because it is in the dependency array
this is a great usecase for `useMemo`

```js
// outside of useEffect!
const courseToUpdate = useMemo(() => {
  coursesList.find((course) => course.id === courseId);
}, [coursesList, courseId]);
```

it is like storing it with useState and updating in a useEffect with [coursesList, courseId] dependency
it does the same thing
(in this case `courseToUpdate` needs to be added to the dep.array of your `useEffect`)

not needed for grading, but I suggest to check dependencies, and extract these kind of things
it is not a big calculation, but still, does not need to run on every dependency change your useEffect currently has

**Review by evasimonyi — APPROVED** [review]
in create course, you should list existing authors, result of fetching from /authors/all
you send this request once but don't do anything with the results, this is why this request is necessary, here should you use the result

2023-09-05_00h07_10



after dispatching the action for add new author in form, Author List should be updated too
from authors list, adding to course authors sections should be the same

- not a requirement, but if user is admin, next to Logout button it still should render something instead of name, like "Admin" -- just a feedback to user, in future you have to keep in mind to always give feedback to user -- not mandatory now
- had some naming-related notes, and some refactor suggestions

great solution, all criteria met, 100% approved
but please check my notes!

File: src/components/CourseForm/CourseForm.jsx (hunk 4 above)
Comment (jurica-zecevic) [reply src/components/CourseForm/CourseForm.jsx:42]: Tried useMemo. Data is not filled in when editing course..

File: src/components/CourseForm/CourseForm.jsx (hunk 3 above)
Comment (jurica-zecevic) [reply src/components/CourseForm/CourseForm.jsx:19]: yep, will do.

File: src/components/Courses/Courses.jsx (hunk 2 above)
Comment (jurica-zecevic) [reply src/components/Courses/Courses.jsx:82]: Changed

File: src/components/CourseForm/CourseForm.jsx (hunk 4 above)
Comment (evasimonyi) [reply src/components/CourseForm/CourseForm.jsx:45]: I might have missed a dependency or something, this should work, but I'll check later. you can merge your current solution and maybe refactor later, this is low prio, just a suggestion
